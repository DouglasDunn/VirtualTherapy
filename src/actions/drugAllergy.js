import database from '../firebase/firebase';

// ADD_DRUG_ALLERGY
export const addDrugAllergy = (drugAllergy) => ({
  type: 'ADD_DRUG_ALLERGY',
  drugAllergy
});

export const createDrugAllergy = (drugAllergies) => {
  return (dispatch) => {
    drugAllergies = drugAllergies.split(',').map(drugAllergy => drugAllergy.trim());
    dispatch(addDrugAllergy(drugAllergies));
  }
};

export const startAddDrugAllergy = (drugAllergyData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      allergicToDrugs = '',
      drugAllergyList = ''
    } = drugAllergyData;
    const drugAllergy = { allergicToDrugs, drugAllergyList };

    return database.ref(`users/${uid}/drugAllergy`).set(drugAllergy).then((ref) => {
      dispatch(addDrugAllergy({
        ...drugAllergy
      }));
    });
  };
};

// SET_DRUG_ALLERGY
export const setDrugAllergy = (drugAllergy) => ({
  type: 'SET_DRUG_ALLERGY',
  drugAllergy
});

export const startSetDrugAllergy = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/drugAllergy`).once('value').then((snapshot) => {
      dispatch(setDrugAllergy(snapshot.val()));
    });
  };
};
