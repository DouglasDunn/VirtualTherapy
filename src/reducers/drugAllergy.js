// DrugAllergy Reducer

const drugAllergyReducerDefaultState = [];

export default (state = drugAllergyReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_DRUG_ALLERGY':
      return [
        ...state,
        ...action.drugAllergy
      ];
    // case 'EDIT_PROFILE':
    //   return {
    //     ...state,
    //     ...action.updates
    //   };
    case 'SET_DRUG_ALLERGY':
      return action.drugAllergy;
    default:
      return state;
  }
};

// case 'ADD_DRUG_ALLERGY':
//   return {
//     ...action.drugAllergy
//   };
