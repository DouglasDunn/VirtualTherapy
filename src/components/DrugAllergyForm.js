import React from 'react';
import { connect } from 'react-redux';
import { addDrugAllergy, createDrugAllergy } from '../actions/drugAllergy';

const initialErrorState = {
  allergicToDrugsError: '',
  drugAllergyListError: ''
};

export class DrugAllergyForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      allergicToDrugs: this.props.drugAllergy.length ? 'yes' : 'no',
      drugAllergyList: this.props.drugAllergy.length ? this.props.drugAllergy.join(", ") : '',
      ...initialErrorState
    };
  }

  onAllergicToDrugsChange = (e) => {
    const allergicToDrugs = e.target.value;
    this.setState(() => ({ allergicToDrugs }));
  };

  onDrugAllergyListChange = (e) => {
    const drugAllergyList = e.target.value;
    this.setState(() => ({ drugAllergyList }));
  };

  validate = () => {
    let allergicToDrugsError = '';
    let drugAllergyListError = '';

    if (!this.state.allergicToDrugs) {
      allergicToDrugsError = 'Please choose an option';
    }

    if ((this.state.allergicToDrugs === 'yes') && !this.state.drugAllergyList) {
      drugAllergyListError = 'drug allergy list cannot be blank';
    } else {
      drugAllergyListError = '';
    }

    if (allergicToDrugsError || drugAllergyListError) {
      this.setState({ allergicToDrugsError, drugAllergyListError });
      return false;
    }

    return true;
  };

  handlePrevious = () => {
    this.props.history.push('/create-profile');
  };

  onSubmit = (e) => {
    e.preventDefault();
    const isValid = this.validate();
    if (isValid) {
      if (this.state.allergicToDrugs === 'no') {
        this.props.addDrugAllergy([]);
      } else {
        this.props.createDrugAllergy(this.state.drugAllergyList);
      }
      this.props.history.push('/add-diagnoses');
    }
  };

  renderDrugAllergyList = () => {
    if (this.state.allergicToDrugs === 'yes') {
      return (
        <div>
          {this.state.drugAllergyListError && <p className="form__error">{this.state.drugAllergyListError}</p>}
          <textarea
            placeholder="Type in all the drugs you are allergic to separated by a comma"
            className="textarea"
            value={this.state.drugAllergyList}
            onChange={this.onDrugAllergyListChange}
          >
          </textarea>
        </div>
      );
    }
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Drug Allergies</h1>
          </div>
        </div>
        <div className="content-container">
          <form className="form">
            {this.state.allergicToDrugsError && <p className="form__error">{this.state.allergicToDrugsError}</p>}
            Are you allergic to any drugs?
            <label>
              <input
                type="radio"
                checked={this.state.allergicToDrugs === 'yes'}
                className="radio-input"
                value="yes"
                onChange={this.onAllergicToDrugsChange}
              />
              Yes
            </label>
            <label>
              <input
                type="radio"
                checked={this.state.allergicToDrugs === 'no'}
                className="radio-input"
                value="no"
                onChange={this.onAllergicToDrugsChange}
              />
              No
            </label>
            {this.renderDrugAllergyList()}
          </form>
          <div>
            <button onClick={this.handlePrevious} className="button">Previous</button>
            <button onClick={this.onSubmit} className="button">Next</button>
          </div>
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  drugAllergy: state.drugAllergy
});

const mapDispatchToProps = (dispatch) => ({
  addDrugAllergy: (drugAllergy) => dispatch(addDrugAllergy(drugAllergy)),
  createDrugAllergy: (drugAllergy) => dispatch(createDrugAllergy(drugAllergy))
});

export default connect(mapStateToProps, mapDispatchToProps)(DrugAllergyForm);
