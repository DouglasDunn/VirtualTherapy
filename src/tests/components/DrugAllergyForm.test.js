import React from 'react';
import { shallow } from 'enzyme';
import DrugAllergyForm from '../../components/DrugAllergyForm';
import drugAllergy from '../fixtures/drugAllergy';

test('should render DrugAllergyForm correctly', () => {
  const wrapper = shallow(<DrugAllergyForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render DrugAllergyForm correctly with drugAllergy data', () => {
  const wrapper = shallow(<DrugAllergyForm drugAllergy={drugAllergy} />);
  expect(wrapper).toMatchSnapshot();
});

// figure out how to test yes and no radio buttons inputs

test('should set drugAllergyList on textarea change', () => {
  const value = 'New drugAllergyList value';
  const wrapper = shallow(<DrugAllergyForm />);
  wrapper.find('textarea').simulate('change', {
    target: { value }
  });
  expect(wrapper.state('drugAllergyList')).toBe(value);
});
