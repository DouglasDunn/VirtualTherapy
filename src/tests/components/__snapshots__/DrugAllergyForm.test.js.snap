// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DrugAllergyForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  Are you allergic to any drugs?
  <label>
    <input
      checked={false}
      className="radio-input"
      onChange={[Function]}
      type="radio"
      value="yes"
    />
    Yes
  </label>
  <label>
    <input
      checked={false}
      className="radio-input"
      onChange={[Function]}
      type="radio"
      value="no"
    />
    No
  </label>
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Type in all the drugs you are allergic to separated by a comma"
    value=""
  />
  <div>
    <button
      className="button"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`should render DrugAllergyForm correctly with drugAllergy data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  Are you allergic to any drugs?
  <label>
    <input
      checked={true}
      className="radio-input"
      onChange={[Function]}
      type="radio"
      value="yes"
    />
    Yes
  </label>
  <label>
    <input
      checked={false}
      className="radio-input"
      onChange={[Function]}
      type="radio"
      value="no"
    />
    No
  </label>
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Type in all the drugs you are allergic to separated by a comma"
    value="Tylenol, Advil, Peptobismol"
  />
  <div>
    <button
      className="button"
    >
      Submit
    </button>
  </div>
</form>
`;
